### ISSUE TITLE
pyrdp.mitm.connections.tcp - unsupported operand type(s) for &: 'NoneType' and 'NegotiationRequestFlags'

### ISSUE THREAD

**Issue by supaplextor** [GoSecure/pyrdp#152]
[Joel880412.pcap.zip](https://github.com/GoSecure/pyrdp/files/3649968/Joel880412.pcap.zip) with attempts before and after VM rebooted (and booted with a new IP)
Ubuntu 18.04.3 LTS
Some client connection attempts trigger the ERRORs below:

```
[2019-09-24 16:46:09,768] - INFO - Joel880412 - pyrdp.mitm.connections.tcp - New client connected from x.x.73.3
[2019-09-24 16:46:09,769] - ERROR - Joel880412 - pyrdp.mitm.connections.tcp - unsupported operand type(s) for &: 'NoneType' and 'NegotiationRequestFlags'
Traceback (most recent call last):
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/tcp.py", line 91, in dataReceived
    self.recv(data)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 143, in recv
    self.pduReceived(pdu)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 198, in pduReceived
    self.next.recv(pdu.payload)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/segmentation.py", line 79, in recv
    layer.recv(forwarded)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/buffered.py", line 55, in recv
    self.pduReceived(pdu)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 198, in pduReceived
    self.next.recv(pdu.payload)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 143, in recv
    self.pduReceived(pdu)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 195, in pduReceived
    Layer.pduReceived(self, pdu)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 110, in pduReceived
    self.observer.onPDUReceived(pdu)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/core/observer.py", line 82, in __call__
    self.composite.doCall(self.item, args, kwargs)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/core/observer.py", line 56, in doCall
    getattr(observer, item)(*args, **kwargs)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 48, in onPDUReceived
    self.handlers[pdu.header](pdu)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/mitm/X224MITM.py", line 62, in onConnectionRequest
    if self.originalRequest.flags & NegotiationRequestFlags.RESTRICTED_ADMIN_MODE_REQUIRED:
TypeError: unsupported operand type(s) for &: 'NoneType' and 'NegotiationRequestFlags'
[2019-09-24 16:46:09,771] - ERROR - Joel880412 - pyrdp.mitm.connections.tcp - Exception occurred when receiving: 0300001f1ae000000000
00436f6f6b69653a206d737473686173683d610d0a

Traceback (most recent call last):
  File "/home/supaplex/Projects/rdp-bait/pyrdp/venv/lib/python3.6/site-packages/twisted/python/log.py", line 103, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/venv/lib/python3.6/site-packages/twisted/python/log.py", line 86, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/venv/lib/python3.6/site-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/venv/lib/python3.6/site-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "/home/supaplex/Projects/rdp-bait/pyrdp/venv/lib/python3.6/site-packages/twisted/internet/asyncioreactor.py", line 136, in _readOrWrite
    why = method()
  File "/home/supaplex/Projects/rdp-bait/pyrdp/venv/lib/python3.6/site-packages/twisted/internet/tcp.py", line 243, in doRead
    return self._dataReceived(data)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/venv/lib/python3.6/site-packages/twisted/internet/tcp.py", line 249, in _dataReceived
    rval = self.protocol.dataReceived(data)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/tcp.py", line 91, in dataReceived
    self.recv(data)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 143, in recv
    self.pduReceived(pdu)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 198, in pduReceived
    self.next.recv(pdu.payload)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/segmentation.py", line 79, in recv
    layer.recv(forwarded)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/buffered.py", line 55, in recv
    self.pduReceived(pdu)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 198, in pduReceived
    self.next.recv(pdu.payload)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 143, in recv
    self.pduReceived(pdu)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 195, in pduReceived
    Layer.pduReceived(self, pdu)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 110, in pduReceived
    self.observer.onPDUReceived(pdu)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/core/observer.py", line 82, in __call__
    self.composite.doCall(self.item, args, kwargs)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/core/observer.py", line 56, in doCall
    getattr(observer, item)(*args, **kwargs)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/layer/layer.py", line 48, in onPDUReceived
    self.handlers[pdu.header](pdu)
  File "/home/supaplex/Projects/rdp-bait/pyrdp/pyrdp/mitm/X224MITM.py", line 62, in onConnectionRequest
    if self.originalRequest.flags & NegotiationRequestFlags.RESTRICTED_ADMIN_MODE_REQUIRED:
builtins.TypeError: unsupported operand type(s) for &: 'NoneType' and 'NegotiationRequestFlags'

[2019-09-24 16:46:09,776] - INFO - Joel880412 - pyrdp.mitm.connections.tcp - Client connection closed. unsupported operand type(s) for &: 'NoneType' and 'NegotiationRequestFlags'
[2019-09-24 16:46:09,776] - INFO - Joel880412 - pyrdp.mitm.connections.tcp - Connection report: report: 1.0, connectionTime: 0.008038520812988281, totalInput: 0, totalOutput: 0
```

**Comment by Res260** [comment]
Can you please provide more details like what were the clients used, what was the destination server OS etc!

**Comment by supaplextor** [comment]
I'm not sure about the client (random attempt from the internet). Destination server is Windows 7 with all its updates.

**Comment by Res260** [comment]
mstshash=a has been known to belong to `morto`, an old RDP worm. It doesn't respect the RDP protocol, but somehow works with some versions of the windows RDP server, because it is more permissive in its implementation of the RDP protocol. It is normal that PyRDP has trouble to complete the connection sequence with that specific client.

That being said, I remember that this client used to exchange more than one packet, so either the client changed or we introduced a breaking change some time ago. More investigation is required for a good conclusion.

**Comment by juliourena** [comment]
Hi @Res260 
I got the same error, here my details.

I'm using 
RDP Server: Windows 10 1903
MIMT: Ubuntu 18.04.3 LTS

Error:
```
[2019-12-30 17:08:09,476] - INFO - Portia610314 - pyrdp.mitm.connections.tcp - New client connected from 170.130.187.10
[2019-12-30 17:08:09,477] - ERROR - Portia610314 - pyrdp.mitm.connections.tcp - unsupported operand type(s) for &: 'NoneType' and 'NegotiationRequestFlags'

Traceback (most recent call last):
  File "/opt/pyrdp/pyrdp/layer/tcp.py", line 91, in dataReceived
    self.recv(data)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 143, in recv
    self.pduReceived(pdu)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 198, in pduReceived
    self.next.recv(pdu.payload)
  File "/opt/pyrdp/pyrdp/layer/segmentation.py", line 79, in recv
    layer.recv(forwarded)
  File "/opt/pyrdp/pyrdp/layer/buffered.py", line 55, in recv
    self.pduReceived(pdu)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 198, in pduReceived
    self.next.recv(pdu.payload)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 143, in recv
    self.pduReceived(pdu)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 195, in pduReceived
    Layer.pduReceived(self, pdu)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 110, in pduReceived
    self.observer.onPDUReceived(pdu)
  File "/opt/pyrdp/pyrdp/core/observer.py", line 82, in __call__
    self.composite.doCall(self.item, args, kwargs)
  File "/opt/pyrdp/pyrdp/core/observer.py", line 56, in doCall
    getattr(observer, item)(*args, **kwargs)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 48, in onPDUReceived
    self.handlers[pdu.header](pdu)
  File "/opt/pyrdp/pyrdp/mitm/X224MITM.py", line 62, in onConnectionRequest
    if self.originalRequest.flags & NegotiationRequestFlags.RESTRICTED_ADMIN_MODE_REQUIRED:
TypeError: unsupported operand type(s) for &: 'NoneType' and 'NegotiationRequestFlags'
[2019-12-30 17:08:09,477] - ERROR - Portia610314 - pyrdp.mitm.connections.tcp - Exception occurred when receiving: 0300002722e00000000000436f6f6b69653a206d737473686173683d63595244674d7048410d
0a

Traceback (most recent call last):
  File "/opt/pyrdp/venv/lib/python3.6/site-packages/twisted/python/log.py", line 103, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/opt/pyrdp/venv/lib/python3.6/site-packages/twisted/python/log.py", line 86, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/opt/pyrdp/venv/lib/python3.6/site-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/opt/pyrdp/venv/lib/python3.6/site-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "/opt/pyrdp/venv/lib/python3.6/site-packages/twisted/internet/asyncioreactor.py", line 136, in _readOrWrite
    why = method()
  File "/opt/pyrdp/venv/lib/python3.6/site-packages/twisted/internet/tcp.py", line 243, in doRead
    return self._dataReceived(data)
  File "/opt/pyrdp/venv/lib/python3.6/site-packages/twisted/internet/tcp.py", line 249, in _dataReceived
    rval = self.protocol.dataReceived(data)
  File "/opt/pyrdp/pyrdp/layer/tcp.py", line 91, in dataReceived
    self.recv(data)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 143, in recv
    self.pduReceived(pdu)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 198, in pduReceived
    self.next.recv(pdu.payload)
  File "/opt/pyrdp/pyrdp/layer/segmentation.py", line 79, in recv
    layer.recv(forwarded)
  File "/opt/pyrdp/pyrdp/layer/buffered.py", line 55, in recv
    self.pduReceived(pdu)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 198, in pduReceived
    self.next.recv(pdu.payload)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 143, in recv
    self.pduReceived(pdu)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 195, in pduReceived
    Layer.pduReceived(self, pdu)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 110, in pduReceived
    self.observer.onPDUReceived(pdu)
  File "/opt/pyrdp/pyrdp/core/observer.py", line 82, in __call__
    self.composite.doCall(self.item, args, kwargs)
  File "/opt/pyrdp/pyrdp/core/observer.py", line 56, in doCall
    getattr(observer, item)(*args, **kwargs)
  File "/opt/pyrdp/pyrdp/layer/layer.py", line 48, in onPDUReceived
    self.handlers[pdu.header](pdu)
  File "/opt/pyrdp/pyrdp/mitm/X224MITM.py", line 62, in onConnectionRequest
    if self.originalRequest.flags & NegotiationRequestFlags.RESTRICTED_ADMIN_MODE_REQUIRED:
builtins.TypeError: unsupported operand type(s) for &: 'NoneType' and 'NegotiationRequestFlags'

[2019-12-30 17:08:09,478] - INFO - Portia610314 - pyrdp.mitm.connections.tcp - Client connection closed. unsupported operand type(s) for &: 'NoneType' and 'NegotiationRequestFlags'
[2019-12-30 17:08:09,478] - INFO - Portia610314 - pyrdp.mitm.connections.tcp - Connection report: report: 1.0, connectionTime: 0.001911163330078125, totalInput: 0, totalOutput: 0


```

**Comment by Res260** [comment]
I am unsure why I made that previous comment, since it seems unrelated to the problem in this issue.

This problem is due to the fact that the first packet sent to the connection is *not* a valid RDP packet. 

It does not influence the outcome of the subsequent connections nor does it make the current connection fail (well, technically, it makes the connection reset, but since the client did not send an RDP packet, its not the direct cause of the failure).

PyRDP wrongly assumes that the first packet of the connection will be a valid RDP packet. It should definitely show a clearer error message instead of this stack trace. However, it does not *break* anything per se because the connection would also fail if it connected directly to the windows machine.

@juliourena @supaplextor

**Comment by alxbl** [comment]
We should handle top-level protocol parsing errors and disconnect clients that send invalid messages. A log message should record the fact that invalid/unparsable protocol data was received, and a debug level logger should probably dump the callstack in case it becomes necessary to debug.

This kind of exception could happen at any point in the parsing with any number of causes. PyRDP trusts that the wire data is sane in almost all of its parsing codebase.

**Comment by obilodeau** [comment]
Just experienced a similar crash with latest master docker image at the same place with different data:

```
[2020-07-29 18:02:53,913] - ERROR - Carolyn484898 - pyrdp.mitm.connections.tcp - unsupported operand type(s) for &: 'NoneType' and 'NegotiationRequestFlags'
Traceback (most recent call last):
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/tcp.py", line 91, in dataReceived
    self.recv(data)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 143, in recv
    self.pduReceived(pdu)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 198, in pduReceived
    self.next.recv(pdu.payload)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/segmentation.py", line 79, in recv
    layer.recv(forwarded)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/buffered.py", line 55, in recv
    self.pduReceived(pdu)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 198, in pduReceived
    self.next.recv(pdu.payload)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 143, in recv
    self.pduReceived(pdu)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 195, in pduReceived
    Layer.pduReceived(self, pdu)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 110, in pduReceived
    self.observer.onPDUReceived(pdu)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/core/observer.py", line 82, in __call__
    self.composite.doCall(self.item, args, kwargs)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/core/observer.py", line 56, in doCall
    getattr(observer, item)(*args, **kwargs)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/layer/layer.py", line 48, in onPDUReceived
    self.handlers[pdu.header](pdu)
  File "/opt/venv/lib/python3.6/site-packages/pyrdp-0.4.2.dev0-py3.6-linux-x86_64.egg/pyrdp/mitm/X224MITM.py", line 62, in onConnectionRequest
    if self.originalRequest.flags & NegotiationRequestFlags.RESTRICTED_ADMIN_MODE_REQUIRED:
TypeError: unsupported operand type(s) for &: 'NoneType' and 'NegotiationRequestFlags'
[2020-07-29 18:02:53,914] - ERROR - Carolyn484898 - pyrdp.mitm.connections.tcp - Exception occurred when receiving: 0300000b06e00000000000
```

* Client: Remmina 1.4.7 / FreeRDP 2.2.0
* Server Windows Server 2016 Datacenter

**Comment by xshill** [comment]
This should be closed by #285
